File: P9_FewShotLearning/Codigo_Profesor/codigo-p9-fewshot-mnist/practica_9_fewshot_mnist_alumnos.jl
# Práctica 9 - Few-shot Learning con MNIST
# Código base para alumnos (Julia)
#
# Objetivo:
# 1. Entrenar una CNN sin ver el dígito 7.
# 2. Reutilizar la CNN como extractor de características.
# 3. Crear prototipos a partir de pocos ejemplos.
# 4. Clasificar imágenes por distancia al prototipo más cercano.
# 5. Comparar 1-shot frente a 5-shot.
#
# Paquetes necesarios:
# ] add Flux MLDatasets Statistics Random LinearAlgebra StatsBase OneHotArrays MultivariateStats Plots

using Flux
using MLDatasets
using Statistics
using Random
using LinearAlgebra
using StatsBase
using OneHotArrays
using MultivariateStats
using Plots

const SEED = 42
Random.seed!(SEED)

const NOVEL_CLASS = 7
const KNOWN_CLASSES = [0, 1, 2, 3, 4, 5, 6, 8, 9]
const COMPACT_CLASSES = collect(1:length(KNOWN_CLASSES))

"""
Carga MNIST y separa el mundo conocido del dígito nuevo.

El modelo se entrenará sin ver ningún 7. Los 7 del conjunto de entrenamiento
se reservan para simular la aparición de una clase nueva con pocos ejemplos.
"""
function load_mnist_world_without_sevens()
    train_x, train_y = MNIST(:train)[:]
    test_x, test_y = MNIST(:test)[:]

    # MLDatasets devuelve imágenes como 28 x 28 x N.
    # Flux espera tensores para convolución como W x H x C x N.
    X_train = Float32.(reshape(train_x, 28, 28, 1, :)) ./ 255f0
    X_test = Float32.(reshape(test_x, 28, 28, 1, :)) ./ 255f0
    y_train = Int.(train_y)
    y_test = Int.(test_y)

    known_mask = y_train .!= NOVEL_CLASS
    seven_mask = y_train .== NOVEL_CLASS

    X_train_known = X_train[:, :, :, known_mask]
    y_train_known = y_train[known_mask]
    X_train_seven = X_train[:, :, :, seven_mask]
    y_train_seven = y_train[seven_mask]

    return X_train_known, y_train_known, X_train_seven, y_train_seven, X_test, y_test
end

"""
Remapea las etiquetas conocidas de MNIST a 1..9 para entrenar la CNN.

Ejemplo: las clases originales son [0,1,2,3,4,5,6,8,9], pero Flux trabaja
más cómodamente con índices compactos [1,2,3,4,5,6,7,8,9].
"""
function remap_known_labels(y)
    mapping = Dict(label => i for (i, label) in enumerate(KNOWN_CLASSES))
    return [mapping[label] for label in y]
end

"""
Construye una CNN sencilla con una capa de embedding explícita.

La penúltima capa Dense(64) se usará después como extractor de características.
"""
function build_classifier(num_classes::Int)
    model = Chain(
        Conv((3, 3), 1 => 32, relu),
        MaxPool((2, 2)),
        Conv((3, 3), 32 => 64, relu),
        MaxPool((2, 2)),
        Flux.flatten,
        Dense(1600 => 64, relu),      # embedding
        Dense(64 => num_classes)      # logits
    )
    return model
end

"""
Entrena la CNN solo con las clases conocidas.
"""
function train_classifier!(model, X_train_known, y_train_known; epochs=3, batchsize=128, lr=1f-3)
    y_compact = remap_known_labels(y_train_known)
    y_onehot = onehotbatch(y_compact, COMPACT_CLASSES)

    data = Flux.DataLoader((X_train_known, y_onehot), batchsize=batchsize, shuffle=true)
    opt_state = Flux.setup(Adam(lr), model)

    for epoch in 1:epochs
        total_loss = 0f0
        nbatches = 0

        for (xb, yb) in data
            loss, grads = Flux.withgradient(model) do m
                logits = m(xb)
                Flux.logitcrossentropy(logits, yb)
            end
            Flux.update!(opt_state, model, grads[1])
            total_loss += Float32(loss)
            nbatches += 1
        end

        println("Epoch $(epoch) - loss = ", round(total_loss / nbatches, digits=4))
    end
end

"""
Evalúa el clasificador inicial solo sobre las clases conocidas del test.
"""
function evaluate_known_classifier(model, X_test, y_test)
    known_mask = y_test .!= NOVEL_CLASS
    X = X_test[:, :, :, known_mask]
    y_original = y_test[known_mask]
    y_compact = remap_known_labels(y_original)

    logits = model(X)
    pred_compact = onecold(logits, COMPACT_CLASSES)
    return mean(pred_compact .== y_compact)
end

"""
Devuelve un extractor de características eliminando la última capa del clasificador.
"""
function create_feature_extractor(model)
    # TODO: devuelve una Chain sin la última capa Dense de clasificación.
    # Pista: Chain(model.layers[1:end-1]...)
    return nothing
end

"""
Selecciona n ejemplos de una clase concreta.
"""
function sample_class_examples(X, y, class_label::Int, n::Int)
    idx = findall(==(class_label), y)
    selected = sample(idx, n; replace=false)
    return X[:, :, :, selected], y[selected]
end

"""
Construye un episodio N-way K-shot usando los dígitos 0..9.

Support set: pocos ejemplos por clase.
Query set: ejemplos de test para evaluar.
"""
function build_fewshot_episode(X_train_known, y_train_known, X_train_seven, y_train_seven, X_test, y_test; n_shots=5, n_query_per_class=100)
    support_x_parts = Vector{Array{Float32, 4}}()
    support_y_parts = Vector{Vector{Int}}()
    query_x_parts = Vector{Array{Float32, 4}}()
    query_y_parts = Vector{Vector{Int}}()

    for class_label in 0:9
        if class_label == NOVEL_CLASS
            Xs, ys = sample_class_examples(X_train_seven, y_train_seven, class_label, n_shots)
        else
            Xs, ys = sample_class_examples(X_train_known, y_train_known, class_label, n_shots)
        end

        Xq, yq = sample_class_examples(X_test, y_test, class_label, n_query_per_class)

        push!(support_x_parts, Xs)
        push!(support_y_parts, ys)
        push!(query_x_parts, Xq)
        push!(query_y_parts, yq)
    end

    X_support = cat(support_x_parts...; dims=4)
    y_support = vcat(support_y_parts...)
    X_query = cat(query_x_parts...; dims=4)
    y_query = vcat(query_y_parts...)

    return X_support, y_support, X_query, y_query
end

"""
Calcula los prototipos de cada clase como la media de sus embeddings.
"""
function compute_prototypes(feature_extractor, X_support, y_support)
    # TODO:
    # 1. Obtén los embeddings del support set con feature_extractor(X_support).
    # 2. Para cada clase, calcula la media de sus embeddings.
    # 3. Devuelve una matriz de prototipos y las etiquetas asociadas.
    return nothing, nothing
end

"""
Clasifica las imágenes de consulta por distancia euclídea al prototipo más cercano.
"""
function classify_by_nearest_prototype(feature_extractor, X_query, prototypes, prototype_labels)
    # TODO:
    # 1. Obtén los embeddings de X_query.
    # 2. Calcula la distancia euclídea de cada embedding a cada prototipo.
    # 3. Asigna a cada imagen la etiqueta del prototipo más cercano.
    return nothing
end

"""
Ejecuta un episodio few-shot completo.
"""
function run_episode(feature_extractor, X_train_known, y_train_known, X_train_seven, y_train_seven, X_test, y_test; n_shots=5)
    X_support, y_support, X_query, y_query = build_fewshot_episode(
        X_train_known, y_train_known,
        X_train_seven, y_train_seven,
        X_test, y_test;
        n_shots=n_shots,
        n_query_per_class=100,
    )

    prototypes, prototype_labels = compute_prototypes(feature_extractor, X_support, y_support)
    y_pred = classify_by_nearest_prototype(feature_extractor, X_query, prototypes, prototype_labels)
    acc = mean(y_pred .== y_query)

    return acc, y_pred, y_query, X_query, prototypes, prototype_labels
end

function main()
    X_train_known, y_train_known, X_train_seven, y_train_seven, X_test, y_test = load_mnist_world_without_sevens()

    println("Clases conocidas: ", KNOWN_CLASSES)
    println("Clase nueva: ", NOVEL_CLASS)
    println("Ejemplos de entrenamiento sin sietes: ", size(X_train_known, 4))
    println("Sietes reservados para few-shot: ", size(X_train_seven, 4))

    model = build_classifier(length(KNOWN_CLASSES))
    train_classifier!(model, X_train_known, y_train_known; epochs=3)

    known_acc = evaluate_known_classifier(model, X_test, y_test)
    println("Accuracy inicial en clases conocidas: ", round(known_acc, digits=4))

    feature_extractor = create_feature_extractor(model)

    # TODO: Ejecuta y compara los episodios 1-shot y 5-shot.
    # Pista:
    # acc_1shot, _, _, _, _, _ = run_episode(...; n_shots=1)
    # acc_5shot, _, _, _, _, _ = run_episode(...; n_shots=5)

    println("Completa los TODO para obtener los resultados finales.")
end

main()

File: P9_FewShotLearning/Codigo_Profesor/codigo-p9-fewshot-mnist/practica_9_fewshot_mnist_alumnos.py
"""
Práctica 9 - Few-shot Learning con MNIST
Código base para alumnos

Objetivo:
    Simular un escenario few-shot en el que el modelo se entrena sin ver
    la clase 7 y posteriormente intenta reconocerla usando pocos ejemplos
    mediante prototipos en un espacio de embeddings.
"""

import os
os.environ["KERAS_BACKEND"] = "jax"

import random
from dataclasses import dataclass

import matplotlib.pyplot as plt
import numpy as np
from sklearn.decomposition import PCA
from sklearn.metrics import accuracy_score
import keras
from keras import layers, models
from keras.src.backend import random as keras_random


SEED = 42
np.random.seed(SEED)
random.seed(SEED)
keras.utils.set_random_seed(SEED)


@dataclass
class FewShotData:
    X_train_known: np.ndarray
    y_train_known: np.ndarray
    X_test: np.ndarray
    y_test: np.ndarray
    known_classes: np.ndarray
    novel_class: int


def load_mnist_world_without_sevens() -> FewShotData:
    (X_train, y_train), (X_test, y_test) = keras.datasets.mnist.load_data()

    X_train = X_train.astype("float32") / 255.0
    X_test = X_test.astype("float32") / 255.0

    X_train = np.expand_dims(X_train, axis=-1)
    X_test = np.expand_dims(X_test, axis=-1)

    mask_known = y_train != 7
    X_train_known = X_train[mask_known]
    y_train_known = y_train[mask_known]

    known_classes = np.array([0, 1, 2, 3, 4, 5, 6, 8, 9])

    return FewShotData(
        X_train_known=X_train_known,
        y_train_known=y_train_known,
        X_test=X_test,
        y_test=y_test,
        known_classes=known_classes,
        novel_class=7,
    )


def build_classifier(num_classes: int) -> keras.Model:
    model = models.Sequential([
        layers.Conv2D(32, (3, 3), activation="relu", input_shape=(28, 28, 1)),
        layers.MaxPooling2D((2, 2)),
        layers.Conv2D(64, (3, 3), activation="relu"),
        layers.MaxPooling2D((2, 2)),
        layers.Flatten(),
        layers.Dense(64, activation="relu", name="embedding"),
        layers.Dense(num_classes, activation="softmax"),
    ])
    model.compile(
        optimizer="adam",
        loss="sparse_categorical_crossentropy",
        metrics=["accuracy"],
    )
    return model


def remap_known_labels(y: np.ndarray, known_classes: np.ndarray) -> np.ndarray:
    mapping = {orig: i for i, orig in enumerate(known_classes)}
    return np.array([mapping[label] for label in y])


def create_feature_extractor(classifier: keras.Model) -> keras.Model:
    # Build the extractor from the classifier's layers (avoiding .input/.output)
    inp = layers.Input(shape=(28, 28, 1))
    x = inp
    for layer in classifier.layers:
        x = layer(x)
        if layer.name == "embedding":
            break
    return models.Model(inputs=inp, outputs=x)


def sample_support_set(X: np.ndarray, y: np.ndarray, class_label: int, n_shots: int) -> tuple[np.ndarray, np.ndarray]:
    indices = np.where(y == class_label)[0]
    selected = np.random.choice(indices, size=n_shots, replace=False)
    return X[selected], y[selected]


def compute_prototypes(feature_extractor: tf.keras.Model, X_support: np.ndarray, y_support: np.ndarray) -> tuple[np.ndarray, np.ndarray]:
    embeddings = feature_extractor.predict(X_support, verbose=0)

    unique_classes = np.unique(y_support)
    prototypes = []
    prototype_labels = []

    for cls in unique_classes:
        mask = y_support == cls
        proto = embeddings[mask].mean(axis=0)
        prototypes.append(proto)
        prototype_labels.append(cls)

    return np.array(prototypes), np.array(prototype_labels)


def classify_by_nearest_prototype(feature_extractor: tf.keras.Model, X_query: np.ndarray, prototypes: np.ndarray, prototype_labels: np.ndarray) -> np.ndarray:
    query_embeddings = feature_extractor.predict(X_query, verbose=0)

    distances = np.linalg.norm(query_embeddings[:, np.newaxis, :] - prototypes[np.newaxis, :, :], axis=2)
    nearest = np.argmin(distances, axis=1)

    return prototype_labels[nearest]


def build_fewshot_episode(data: FewShotData, n_shots: int, n_query_per_class: int = 100):
    X_support_list, y_support_list = [], []
    X_query_list, y_query_list = [], []

    X_train = data.X_train_known
    y_train = data.y_train_known
    X_test = data.X_test
    y_test = data.y_test

    for cls in range(10):
        if cls in data.known_classes:
            sup_X, sup_y = sample_support_set(X_train, y_train, cls, n_shots)
        else:
            test_indices = np.where(y_test == cls)[0]
            chosen = np.random.choice(test_indices, size=n_shots, replace=False)
            sup_X = X_test[chosen]
            sup_y = y_test[chosen]

        X_support_list.append(sup_X)
        y_support_list.append(sup_y)

        test_indices = np.where(y_test == cls)[0]
        qry_indices = np.random.choice(test_indices, size=n_query_per_class, replace=False)
        X_query_list.append(X_test[qry_indices])
        y_query_list.append(y_test[qry_indices])

    X_support = np.concatenate(X_support_list, axis=0)
    y_support = np.concatenate(y_support_list, axis=0)
    X_query = np.concatenate(X_query_list, axis=0)
    y_query = np.concatenate(y_query_list, axis=0)

    return X_support, y_support, X_query, y_query


def plot_accuracy_comparison(results: dict[str, float], output_path: str = "fewshot_accuracy_comparison.png") -> None:
    plt.figure(figsize=(6, 4))
    bars = plt.bar(list(results.keys()), list(results.values()), color=["#ff7f0e", "#1f77b4"])
    for bar, val in zip(bars, results.values()):
        plt.text(bar.get_x() + bar.get_width() / 2, bar.get_height() + 0.01,
                 f"{val:.4f}", ha="center", va="bottom", fontsize=11)
    plt.ylim(0, 1)
    plt.ylabel("Accuracy")
    plt.title("Few-shot classification: 1-shot vs 5-shot")
    plt.grid(axis="y", alpha=0.3)
    plt.tight_layout()
    plt.savefig(output_path, dpi=200)
    plt.show()


def plot_embeddings_pca(feature_extractor: tf.keras.Model, X: np.ndarray, y: np.ndarray, prototypes: np.ndarray | None = None, prototype_labels: np.ndarray | None = None, output_path: str = "fewshot_embeddings_pca.png") -> None:
    embeddings = feature_extractor.predict(X, verbose=0)

    pca = PCA(n_components=2)
    emb_2d = pca.fit_transform(embeddings)

    plt.figure(figsize=(8, 6))
    scatter = plt.scatter(emb_2d[:, 0], emb_2d[:, 1], c=y, cmap="tab10", alpha=0.6, s=10)

    if prototypes is not None:
        proto_2d = pca.transform(prototypes)
        plt.scatter(proto_2d[:, 0], proto_2d[:, 1], c=prototype_labels, cmap="tab10",
                    marker="X", s=200, edgecolors="black", linewidths=1.5, label="Prototypes")

    plt.colorbar(scatter, label="Digit class")
    plt.title("2D PCA of MNIST embeddings (few-shot feature space)")
    plt.xlabel(f"PC1 ({pca.explained_variance_ratio_[0]:.1%} variance)")
    plt.ylabel(f"PC2 ({pca.explained_variance_ratio_[1]:.1%} variance)")
    plt.tight_layout()
    plt.savefig(output_path, dpi=200)
    plt.show()


def main() -> None:
    data = load_mnist_world_without_sevens()

    y_train_known_remap = remap_known_labels(data.y_train_known, data.known_classes)
    classifier = build_classifier(num_classes=len(data.known_classes))

    classifier.fit(
        data.X_train_known,
        y_train_known_remap,
        validation_split=0.1,
        epochs=3,
        batch_size=128,
        verbose=1,
    )

    feature_extractor = create_feature_extractor(classifier)

    sample_emb = feature_extractor.predict(data.X_test[:1], verbose=0)
    print(f"Embedding dimension: {sample_emb.shape[-1]}")

    results = {}
    for n_shots in [1, 5]:
        X_support, y_support, X_query, y_query = build_fewshot_episode(data, n_shots=n_shots)
        prototypes, prototype_labels = compute_prototypes(feature_extractor, X_support, y_support)
        y_pred = classify_by_nearest_prototype(feature_extractor, X_query, prototypes, prototype_labels)
        acc = accuracy_score(y_query, y_pred)
        results[f"{n_shots}-shot"] = acc
        print(f"{n_shots}-shot accuracy: {acc:.4f}")

    plot_accuracy_comparison(results)

    # Build a combined query set for PCA visualization
    _, _, viz_X, viz_y = build_fewshot_episode(data, n_shots=1, n_query_per_class=50)
    # Build a support set to mark prototype positions on the PCA plot
    sup_X, sup_y, _, _ = build_fewshot_episode(data, n_shots=5, n_query_per_class=0)
    viz_protos, viz_proto_labels = compute_prototypes(feature_extractor, sup_X, sup_y)
    plot_embeddings_pca(feature_extractor, viz_X, viz_y, prototypes=viz_protos, prototype_labels=viz_proto_labels)


if __name__ == "__main__":
    main()
